lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,662 @@
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Depth-first walk through the DOM tree.
+     *
+     * @param DOMNode $node Node to start walking from.
+     * @return DOMNode|null Next node, or null if none found.
+     */
+    private function nextNode(DOMNode $node)
+    {
+        // Walk downwards if there are children.
+        if ($node->firstChild) {
+            return $node->firstChild;
+        }
+
+        // Return direct sibling or walk upwards until we find a node with a sibling.
+        while ($node) {
+            if ($node->nextSibling) {
+                return $node->nextSibling;
+            }
+
+            $node = $node->parentNode;
+        }
+
+        // Out of nodes, so we're done.
+        return null;
+    }
+
+    /**
+     * Skip the subtree that is descending from the provided node.
+     *
+     * @param DOMNode $node Node to skip the subtree of.
+     * @return DOMNode|null The appropriate "next" node that will skip the current subtree, null if none found.
+     */
+    private function skipNodeAndChildren(DOMNode $node)
+    {
+        if ($node->nextSibling) {
+            return $node->nextSibling;
+        }
+
+        return $this->skipNodeAndChildren($node->parentNode);
+    }
+
+    /**
+     * Check whether a given element is an AMP embed.
+     *
+     * @param DOMElement $element Element to check.
+     * @return bool Whether the given element is an AMP embed.
+     */
+    private function isAmpEmbed(DOMElement $element)
+    {
+        return array_key_exists($element->tagName, self::AMP_EMBEDS);
+    }
+
+    /**
+     * Get the URL of the CSS background-image property.
+     *
+     * This falls back to the data-amp-original-style attribute if the inline
+     * style was already extracted by the CSS tree-shaking.
+     *
+     * @param DOMElement $element
+     * @return string URL of the background image, or an empty string if not found.
+     */
+    private function getCssBackgroundImageUrl(DOMElement $element)
+    {
+        $matches = [];
+
+        if (
+            preg_match(
+                self::CSS_BACKGROUND_IMAGE_URL_REGEX_PATTERN,
+                $element->getAttribute(Attribute::STYLE),
+                $matches
+            )
+        ) {
+            return trim($matches['url'], '\'" ');
+        }
+
+        if (
+            !empty($this->inlineStyleBackupAttribute)
+            && preg_match(
+                self::CSS_BACKGROUND_IMAGE_URL_REGEX_PATTERN,
+                $element->getAttribute($this->inlineStyleBackupAttribute),
+                $matches
+            )
+        ) {
+            return trim($matches['url'], '\'" ');
+        }
+
+        return '';
+    }
+
+    /**
+     * Whether srcset preloading is supported.
+     *
+     * @return bool
+     */
+    private function supportsSrcset()
+    {
+        // TODO: Add logic here or override manually.
```

Comment:
This could pass through a `PRELOAD_SRCSET` configuration option that defaults to `false`, yeah?